### PR TITLE
Implement Rubocop

### PR DESCRIPTION
**Pull request by renburnett**
[//]: # (Title Template: "[TR_##] Title")
# Implement Rubocop

#### Please check if the PR fulfills these requirements

- [x] The changes don't come from your master branch. ([Source](https://blog.jasonmeridth.com/posts/do-not-issue-pull-requests-from-your-master-branch/)) 

---

#### What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)

- setup and modify rubocop config file
- add linting guidelines in README.md
- fix rubocop layout/syntax errors 
- refactor several methods in accordance with ruby/rails style guide where possible
- temporarily disable some cops to avoid slightly larger refactor of a few methods
  - add TODOs to methods requiring larger refactors

#### What is the current behavior? (You can also link to an open issue here)
- **[github issue](https://github.com/FoodIsLifeBGP/banana-rails/issues/48)**

#### What is the new behavior? (if this is a feature change)
- add a rubocop linter to standardize codebase

#### Does this PR introduce a breaking change? (What changes might users need to make in their application due to this PR?)
- not that I'm aware of 👍 

#### Other information
- please feel free to add any comments, suggestions or changes that you think should be added!

#### Discussion Questions
- should we implement the [Style/FrozenStringLiteral](https://www.rubydoc.info/gems/rubocop/RuboCop/Cop/Style/FrozenStringLiteralComment) cop?
  - it adds a pragma to the top of pretty much every single ruby file
  - frozen string literals may be default in the next ruby version
- `get_donations` and `get_claims` in `clients_controller.rb` fail the following rubocop checks: 
    - [Metrics/AbcSize](https://rubocop.readthedocs.io/en/latest/cops_metrics/#metricsabcsize), [Lint/UselessAssignment](https://rubocop.readthedocs.io/en/latest/cops_lint/#lintuselessassignment), [Rails/Output](https://docs.rubocop.org/projects/rails/en/stable/cops_rails/#railsoutput), [Metrics/CyclomaticComplexity](https://rubocop.readthedocs.io/en/latest/cops_metrics/#metricscyclomaticcomplexity), [Metrics/MethodLength](https://rubocop.readthedocs.io/en/latest/cops_metrics/#metricsmethodlength); some of these errors will be fixed when the app goes into production, however big chunks of these methods will still need a bit of refactoring at that time. Thoughts? Leave as-is?
- `make_claim` in `donations_controller.rb` fails the following rubocop checks: 
    - Metrics/MethodLength, Metrics/AbcSize
- [**google doc**](https://docs.google.com/spreadsheets/d/1Hmox0ZP7nBOiBhULHWL_bSeFcMLsTMen-maH1Cqz130/edit?usp=sharing)

---

#### TODO: 
- Decide whether or not to refactor `get_donations`, `get_claims` and/or `make_claim`

### CODE REVIEW

File: db/schema.rb [hunk 1]
```diff
@@ -44,11 +44,12 @@
     t.string "address_street"
     t.string "address_city"
     t.string "address_state"
-    t.integer "address_zip"
+    t.string "address_zip"
```
Comment (rahulrajaram) [inline db/schema.rb:47]: Why did this change? Does this not break migrations?

File: db/schema.rb [hunk 2]
```diff
@@ -81,12 +82,13 @@
...
     t.string "business_license"
     t.string "account_status"
     t.string "pickup_location"
     t.decimal "latitude", precision: 10, scale: 6
     t.decimal "longitude", precision: 10, scale: 6
+    t.index ["email"], name: "index_donors_on_email", unique: true
```
Comment (rahulrajaram) [inline db/schema.rb:91]: Did Rubocop report this? Either way, this change is a sneaky one and doesn't belong in this PR.

File: db/seeds.rb [hunk 3]
```diff
@@ -395,60 +394,58 @@
...
-		completed: completed,
-		time_claimed: Time.now,
-		canceled: false
-	)
+  Claim.create!(
+    client_id: client_id,
+    donation_id: donation_id,
+    qr_code: Base64.encode64({ 'client_id': client_id, 'donation_id': donation_id }.to_json).chomp,
+    completed: completed,
+    time_claimed: Time.zone.now,
```
Comment (rahulrajaram) [inline db/seeds.rb:446]: How does Rails save `Time.zone.now` in the DB? Does it store the UTC time or local time? We should prefer UTC.

File: db/seeds.rb [hunk 4]
```diff
@@ -467,11 +464,9 @@ def create_claim(client_id, donation_id, completed)
...
 claims.each { |c| create_claim(*c) }
-
-puts "Seeding Admins..."
 Admin.create!(
 	email: 'admin@admin.com',
 	password: 'admin',
 )
 
-puts "Finished with no errors"
+logger.debug('Seeding Admin...')
```
Comment (rahulrajaram) [inline db/seeds.rb:472]: Missing newline at the end of file.

File: db/schema.rb [hunk 5]
```diff
@@ -10,7 +10,7 @@
 #
 # It's strongly recommended that you check this file into your version control system.
 
-ActiveRecord::Schema.define(version: 2019_12_13_210446) do
+ActiveRecord::Schema.define(version: 2020_04_22_064207) do
```
Comment (rahulrajaram) [inline db/schema.rb:13]: You must not change schema versions. Given a schema version all DB migrations prior to this date will be assumed to have run.

File: db/schema.rb (hunk 1 above)
Comment (renburnett) [reply db/schema.rb:47]: I updated the type of `address_zip` to match the frontend type, there shouldn't be a disparity between the two.

Comment (renburnett) [reply db/schema.rb:47]: Also I don't know if you're aware but when you add a migration the schema changes.

File: db/seeds.rb (hunk 4 above)
Comment (renburnett) [reply db/seeds.rb:472]: Thanks

File: db/seeds.rb (hunk 3 above)
Comment (renburnett) [reply db/seeds.rb:446]: if you look at `config/application.rb` you'll see where the timezone is set. I set it to PST but I'm fine changing it to UTC

File: db/schema.rb (hunk 2 above)
Comment (renburnett) [reply db/schema.rb:91]: In Rails when you add a migration, the schema changes (please see the `db/migrations` that are included on this PR). I added a `unique` column to prevent race condition. I've linked the relevant article and the Cop that flagged this code in case other reviewers have similar questions:
- [the cop](https://docs.rubocop.org/projects/rails/en/stable/cops_rails/#railsuniquevalidationwithoutindex)
- [the issue](https://thoughtbot.com/blog/the-perils-of-uniqueness-validations)

File: db/schema.rb (hunk 5 above)
Comment (renburnett) [reply db/schema.rb:13]: The schema updated when I ran `rails db:migrate` with the new migrations that I wrote/added. My understanding is that this is acceptable? However I am happy to change this if I'm incorrect. Perhaps we could sync offline or could you link an article for me? Thanks!

File: db/schema.rb (hunk 1 above)
Comment (sagehen03) [reply db/schema.rb:47]: I think this change is also a good idea from the standpoint of supporting zips that start with 0.  Probably won't ever be necessary, but just in case.

**Comment by emarti43** [comment]
Gonna close this. New PR is up.
